artichoke-backend/src/extn/core/time/subsec.rs

```diff
@@ -0,0 +1,562 @@
+            match subsec_symbol {
+                b"milliseconds" => Ok(SubsecMultiplier::Millis),
+                b"usec" => Ok(SubsecMultiplier::Micros),
+                b"nsec" => Ok(SubsecMultiplier::Nanos),
+                _ => {
+                    let mut message = b"unexpected unit: ".to_vec();
+                    message.extend_from_slice(subsec_symbol);
+                    Err(ArgumentError::from(message).into())
+                }
+            }
+        } else {
+            Ok(SubsecMultiplier::Micros)
+        }
+    }
+}
+
+/// A struct that represents the adjustment needed to a `Time` based on a
+/// the parsing of optional Ruby Values. Seconds can require adjustment as a
+/// means for handling overflow of values. e.g. `1_001` millis can be requested
+/// which should result in 1 seconds, and `1_000_000` nanoseconds.
+///
+/// Note: Negative nanoseconds are not supported, thus any negative adjustment
+/// will generally result in at least -1 second, and the relevant positive
+/// amount of nanoseconds. e.g. `-1_000` microseconds should result in -1
+/// second, and `999_999_000` nanoseconds.
+#[derive(Debug, Copy, Clone)]
+pub struct Subsec {
+    secs: i64,
+    nanos: u32,
+}
+
+impl Subsec {
+    /// Returns a tuple of (seconds, nanoseconds). Subseconds are provided in
+    /// various accuracies, and can overflow. e.g. 1001 milliseconds, is 1
+    /// second, and `1_000_000` nanoseconds.
+    #[must_use]
+    pub fn to_tuple(&self) -> (i64, u32) {
+        (self.secs, self.nanos)
+    }
+}
+
+impl TryConvertMut<(Option<Value>, Option<Value>), Subsec> for Artichoke {
+    type Error = Error;
+
+    fn try_convert_mut(&mut self, params: (Option<Value>, Option<Value>)) -> Result<Subsec, Self::Error> {
+        let (subsec, subsec_unit) = params;
+
+        if let Some(subsec) = subsec {
+            let multiplier: SubsecMultiplier = self.try_convert_mut(subsec_unit)?;
+            let multiplier_nanos = multiplier.as_nanos();
+            // `subsec` represents the user provided value in `subsec_unit`
+            // resolution. The base used to derive the number of seconds is
+            // based on the `subsec_unit`. e.g. `1_001` milliseconds is 1
+            // second, and `1_000_000` nanoseconds.
+            let seconds_base = NANOS_IN_SECOND / multiplier_nanos;
+
+            if subsec.ruby_type() == Ruby::Float {
+                // FIXME: The below deviates from the MRI implementation of
+                // Time. MRI uses `to_r` for subsec calculation on floats
+                // subsec nanos, and this could result in different values.
+
+                let subsec: f64 = self.try_convert(subsec)?;
+
+                if subsec.is_nan() {
+                    return Err(FloatDomainError::with_message("NaN").into());
+                }
+                if subsec.is_infinite() {
+                    return Err(FloatDomainError::with_message("Infinity").into());
```

Comment:
negative infinity gets a different exception message:

```
[3.1.2] > Time.at(12, -Float::INFINITY, :usec)
<internal:timev>:274:in `to_r': -Infinity (FloatDomainError)
```

Reply:
also fun, it looks like this converts through `Rational` which Artichoke doesn't support yet.

Reply:
added a test and added a branch for `subsec.is_sign_negative()`.